### PR TITLE
[KBFS-1641] Add more logging to help track down races

### PR DESCRIPTION
**Pull request by akalin-keybase**
Make kbfsblock, kbfscrypto, kbfshash packages, and
tlf.ID use pkg/errors to wrap errors. Annotate READMEs
with which packages (or portions) use pkg/errors.

Also make more tests in the packages above use
require/assert.

Make fewer functions use implicit returns.

Move some crypto tests from libkbfs to kbfscrypto.

Print errors out with "%+v" in more places, in particular,
key_manager_test.go.

Attach original errors to Need{Self,Other}RekeyError.

### CODE REVIEW

File: kbfsblock/id_test.go [hunk 1]
```diff
@@ -19,52 +19,44 @@ func TestIDEncodeDecode(t *testing.T) {
...
-	if err != nil {
-		t.Fatal(err)
-	}
+	require.NoError(t, err)
+	require.Equal(t, ID{}, id)
+}
+
+// Test (very superficially) that MakeTemporaryID() returns non-zero
+// values that aren't equal.
+func TestTemporaryIDRandom(t *testing.T) {
```
Comment (akalin-keybase) [inline kbfsblock/id_test.go]: Moved this test from libkbfs

File: kbfscrypto/crypto_key_types_test.go [hunk 2]
```diff
@@ -285,26 +284,44 @@ func TestBlockCryptKeyEncodeDecode(t *testing.T) {
...
-	if k2 == (BlockCryptKeyServerHalf{}) {
-		t.Errorf("zero BlockCryptKeyServerHalf k2")
-	}
+	require.NotEqual(t, k1, k2)
+}
+
+// Test that MaskTLFCryptKey() returns bytes that are different from
+// the server half and the key, and that UnmaskTLFCryptKey() undoes
+// the masking properly.
+func TestMaskUnmaskTLFCryptKey(t *testing.T) {
```
Comment (akalin-keybase) [inline kbfscrypto/crypto_key_types_test.go]: Moved these tests from libkbfs

File: kbfscrypto/crypto_key_types.go [hunk 3]
```diff
@@ -329,6 +337,19 @@ func MakeTLFCryptKeyServerHalf(data [32]byte) TLFCryptKeyServerHalf {
 	return TLFCryptKeyServerHalf{publicByte32Container{data}}
 }
 
+// MakeRandomTLFCryptKeyServerHalf generates the server-side of a
+// top-level folder crypt key.
+func MakeRandomTLFCryptKeyServerHalf() (
```
Comment (akalin-keybase) [inline kbfscrypto/crypto_key_types.go]: Moved this fn from libkbfs

File: kbfscrypto/crypto_key_types.go [hunk 4]
```diff
@@ -373,6 +394,16 @@ func MakeTLFCryptKey(data [32]byte) TLFCryptKey {
 	return TLFCryptKey{privateByte32Container{data}}
 }
 
+// MakeRandomTLFCryptKey returns a random top-level folder crypt key.
+func MakeRandomTLFCryptKey() (TLFCryptKey, error) {
```
Comment (akalin-keybase) [inline kbfscrypto/crypto_key_types.go]: Created this new fn (used by `libkbfs.CryptoCommon`)

File: kbfscrypto/signature_test.go [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+import (
+	"testing"
+
+	"github.com/keybase/client/go/libkb"
+	"github.com/pkg/errors"
+	"github.com/stretchr/testify/assert"
+)
+
+// Test that Verify() rejects various types of bad signatures.
+func TestVerifyFailures(t *testing.T) {
```
Comment (akalin-keybase) [inline kbfscrypto/signature_test.go]: Moved these tests from libkbfs

File: libkbfs/crypto_client.go [hunk 6]
```diff
@@ -118,28 +117,37 @@ func (c *CryptoClient) SignToString(ctx context.Context, msg []byte) (
...
-		return
+		return keybase1.EncryptedBytes32{}, keybase1.BoxNonce{},
+			errors.WithStack(UnknownEncryptionVer{
+				encryptedClientHalf.Version})
 	}
 
 	if len(encryptedClientHalf.EncryptedData) != len(encryptedData) {
-		err = libkb.DecryptionError{}
-		return
+		return keybase1.EncryptedBytes32{}, keybase1.BoxNonce{},
```
Comment (akalin-keybase) [inline libkbfs/crypto_client.go]: changed this error to avoid confusion with errors returned from the service

File: libkbfs/crypto_local.go [hunk 7]
```diff
@@ -94,27 +97,34 @@ func (c CryptoLocal) DecryptTLFCryptKeyClientHalfAny(ctx context.Context,
...
 	clientHalf kbfscrypto.TLFCryptKeyClientHalf, index int, err error) {
 	if len(keys) == 0 {
-		return clientHalf, index, NoKeysError{}
+		return kbfscrypto.TLFCryptKeyClientHalf{}, -1,
+			errors.WithStack(NoKeysError{})
 	}
+	var firstNonDecryptionErr error
 	for i, k := range keys {
-		nonce, err := c.prepareTLFCryptKeyClientHalf(k.ClientHalf, clientHalf)
+		clientHalf, err := c.DecryptTLFCryptKeyClientHalf(
```
Comment (akalin-keybase) [inline libkbfs/crypto_local.go]: Rewrote this loop a bit

**Comment by strib** [comment]
giphy

**Pull request merged**